network/noise/src/socket.rs

```diff
@@ -518,7 +552,16 @@ where
     /// (switched to transport mode) upon success.
     pub async fn handshake_1rt(mut self) -> io::Result<NoiseSocket<TSocket>> {
         // The Dialer
-        if self.0.session.is_initiator() {
+        let initiator = if let Session::Handshake(ref session) = self.0.session {
+            session.is_initiator()
+        } else {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Session not in handshake state".to_string(),
+            ));
+        };
```

Comment:
Maybe we can have a method on Session that does this check for us to keep this logic simple here?